File: scraps/gpm-track.c
// Proxy for GPM [ https://github.com/telmich/gpm ] mouse events to shm/signals
// Build with: gcc -O2 gpm-track.c -o gpm-track -lgpm -lrt
// Usage example: ./gpm-track --pid 26629 --shm gpm-test < /dev/tty3
// More info: ./gpm-track -h
// See also: gpm-track.py in the same dir.

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <sys/mman.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <signal.h>

#include <getopt.h>
#include <limits.h>
#include <errno.h>

// Good gpm tutorial/examples - https://www.linuxjournal.com/article/4600
// Docs: info gpm
#include <gpm.h>

#define P(err, fmt, arg...)\
	do {\
		fprintf(stderr, "ERROR: " fmt "\n", ##arg);\
		if (err) exit(err);\
	} while (0)

long shm_len;
void *shm;
int signal_pid = 0, signal_base = 40;

void *create_shared_memory(char *filename, size_t size) {
	void *shm = NULL;

	mode_t umask_mode = umask(0);
	int mmap_fd = shm_open(filename, O_CREAT|O_RDWR, 00600);
	umask(umask_mode);
	if (mmap_fd < 0) { perror("shm_open() failed"); return NULL; }

	int res = ftruncate(mmap_fd, size) == 0 ? 1 : 0;
	if (res) res = lseek(mmap_fd, size - 1, SEEK_SET) != -1 ? 1 : 0;
	if (res) res = write(mmap_fd, "", 1) != -1 ? 1 : 0;
	if (res) {
		shm = mmap( NULL, size,
			PROT_READ | PROT_WRITE, MAP_SHARED, mmap_fd, 0 );
		res = (shm != MAP_FAILED && shm != NULL) ? 1 : 0; }
	if (!res) { perror("ERROR: mmap failed"); close(mmap_fd); return NULL; }

	return shm;
}

int event_handler(Gpm_Event *event, void *data) {
	sprintf((char *) data, "%05d %05d\n", event->x, event->y);
	if (event->type & GPM_DOWN) {
		int click_t =
			event->type & GPM_SINGLE ? 1 :
			event->type & GPM_DOUBLE ? 2 :
			event->type & GPM_TRIPLE ? 3 : 0;
		int button =
			event->buttons & GPM_B_LEFT ? 1 :
			event->buttons & GPM_B_RIGHT ? 2 :
			event->buttons & GPM_B_MIDDLE ? 3 : 0;
		sprintf( ((char *) data) + 12,
			"{\"b\": %d, \"t\": %d, \"x\": %d, \"y\": %d}\n",
			button, click_t, event->x, event->y );
		if (signal_pid && button > 0)
			kill(signal_pid, signal_base + button + (click_t << 2));
	}
	return 0;
}

void parse_opts( int argc, char *argv[],
		char **opt_shm, int *opt_pid ) {
	extern char *optarg;
	extern int optind, opterr, optopt;

	int res = (*opt_shm = malloc(32)) != NULL ? 1 : 0;
	if (res) res = snprintf(*opt_shm, 32, "gpm-track.%d", getpid()) < 32 ? 1 : 0;
	if (!res) P(1, "failed init -s/--shm value");

	void usage(int err) {
		FILE *dst = !err ? stdout : stderr;
		fprintf( dst,
"Usage: %s [-h|--help] [-s|--shm file] [-p|--pid pid] < ttyX\n\n"
"Handle libgpm events and dump info on these"
	" to specified -s/--shm file (default: 'gpm-track.{pid}').\n\n"
"If -p/--pid is specified, it will be sent SIGRT-X on any mouse events,\n"
	" where 'X' is '%d + mask' and 'mask' is (button | 2 << clicks),\n"
	" button={1=left, 2=right, 3=middle},"
	" clicks={1=single, 2=double, 3=triple}.\n\n"
			, argv[0], signal_base );
		exit(err); }

	int ch;
	static struct option opt_list[] = {
		{"help", no_argument, NULL, 1},
		{"shm", required_argument, NULL, 2},
		{"pid", required_argument, NULL, 3} };
	while ((ch = getopt_long(argc, argv, ":hs:p:", opt_list, NULL)) != -1)
		switch (ch) {
			case 's': case 2: *opt_shm = strdup(optarg); break;
			case 'p': case 3:
				errno = 0;
				*opt_pid = (int) strtol(optarg, NULL, 10);
				if (*opt_pid <= 0 || (errno == ERANGE && (
						*opt_pid == (int) LONG_MIN || *opt_pid == (int) LONG_MAX )))
					P(1, "invalid pid value: %s", optarg);
				break;
			case 'h': case 1: usage(0);
			case '?':
				P(0, "unrecognized option - %s\n", argv[optind-1]);
				usage(1);
			case ':':
				if (optopt >= 32) P(0, "missing argument for -%c\n", optopt);
				else P(0, "missing argument for --%s\n", opt_list[optopt-1].name);
				usage(1);
			default: usage(1); }
	if (optind < argc) {
		P(0, "unrecognized argument value - %s\n", argv[optind]);
		usage(1); }
}

int main(int argc, char *argv[]) {
	char *opt_shm; int opt_pid = 0;
	parse_opts(argc, argv, &opt_shm, &opt_pid);
	if (opt_pid > 0) signal_pid = opt_pid;
	if ( SIGRTMIN > signal_base ||
			SIGRTMAX < signal_base + (1 << 4) )
		P( 1,
			"no SIGRT* space: range=[%d - %d], need=[%d - %d]",
			SIGRTMIN, SIGRTMAX, signal_base, signal_base + (1 << 4) );

	shm_len = sysconf(_SC_PAGESIZE);
	shm = create_shared_memory(opt_shm, (size_t) shm_len);
	if (!shm) P(1, "shm failed");

	Gpm_Connect conn; int c;

	conn.eventMask = ~0; // want to know about all the events
	conn.defaultMask = 0; // don't handle anything by default
	conn.minMod = 0; // want everything
	conn.maxMod = ~0; // all modifiers included

	setenv("TERM", "linux", 1); // Gpm_Open returns raw stdin if not linux/xterm
	if (Gpm_Open(&conn, 0) == -1)
		P(1, "failed to connect to gpm server");

	gpm_handler = event_handler;
	gpm_data = shm;

	while ((c = Gpm_Getc(stdin)) != EOF) printf("%c", c);
	Gpm_Close();

	return 0;
}

File: scraps/gpm-track.py
#!/usr/bin/env python3

import os, sys, mmap, time
import signal, subprocess, contextlib, collections

def retries_within_timeout( tries, timeout,
		backoff_func=lambda e,n: ((e**n-1)/e), slack=1e-2 ):
	'Return list of delays to make exactly n tires within timeout, with backoff_func.'
	a, b = 0, timeout
	while True:
		m = (a + b) / 2
		delays = list(backoff_func(m, n) for n in range(tries))
		error = sum(delays) - timeout
		if abs(error) < slack: return delays
		elif error > 0: b = m
		else: a = m

class GPMError(Exception): pass

def main(args=None):
	import argparse
	parser = argparse.ArgumentParser(
		description='Track and print mouse events from gpm.')
	parser.add_argument('tty', help='TTY name without /dev prefix (example: tty3).')
	parser.add_argument('-i', '--interval',
		metavar='seconds', default=1.0, type=float,
		help='Interval between reading x/y position. Default: %(default)s')
	parser.add_argument('-s', '--shm', metavar='name',
		help='/dev/shm name to use for same gpm-track option (empty: do not pass).')
	parser.add_argument('-b', '--gpm-track-binary',
		metavar='path', default='./gpm-track',
		help='Path to compiled gpm-track binary Default: %(default)s.')
	opts = parser.parse_args(sys.argv[1:] if args is None else args)

	cmd = [opts.gpm_track_binary, '--pid', str(os.getpid())]
	if opts.shm: cmd += ['--shm', opts.shm]

	for sig in range(40, 57): signal.signal(sig, signal.SIG_IGN)

	with contextlib.ExitStack() as ctx:
		tty = ctx.enter_context(open(f'/dev/{opts.tty}', 'rb'))
		proc = ctx.enter_context(subprocess.Popen(cmd, stdin=tty))
		ctx.callback(proc.terminate)

		shm = '/dev/shm/{}'.format(opts.shm or f'gpm-track.{proc.pid}')
		for delay in retries_within_timeout(10, 5):
			if os.path.exists(shm): break
			try: err = proc.wait(delay)
			except subprocess.TimeoutExpired: pass
			else: raise GPMError(f'gpm-track pid crashed (code={err})')
		else: raise GPMError(f'gpm-track pid failed to create shm file: {shm}')

		fd = os.open(shm, os.O_RDWR | os.O_SYNC)
		ctx.callback(os.close, fd)
		mem = mmap.mmap( fd, mmap.PAGESIZE,
			mmap.MAP_SHARED, mmap.PROT_READ | mmap.PROT_WRITE, offset=0 )

		pos_t = collections.namedtuple('pos', 'x y')
		def read_pos():
			mem.seek(0)
			pos = mem.read(12)
			if not pos.endswith(b'\n'): return
			return pos_t(*map(int, pos.decode().split()))

		click_report = None
		click_t = collections.namedtuple('click', 't btn pos')
		def read_click( sig, frm,
				_button=['none', 'left', 'right', 'middle'],
				_click_t=['none', 'single', 'double', 'triple'] ):
			nonlocal click_report
			sig -= 40
			click_report = click_t(
				_click_t[(sig >> 2) & 3],
				_button[sig & 3], read_pos() )
		for sig in range(40, 57): signal.signal(sig, read_click)

		pos_last, delay = None, opts.interval
		while True:
			if click_report:
				print('click', click_report)
				click_report = None
			else:
				pos = read_pos()
				if pos != pos_last:
					pos_last = pos
					print('pos:', pos)
			time.sleep(delay)

if __name__ == '__main__': sys.exit(main())
